File: packages/Thoth.Json.Core/Encode.fs
namespace Thoth.Json.Core

open System.Globalization
open System

[<RequireQualifiedAccess>]
module Encode =

    /// <summary>Encode a string.</summary>
    let inline string value =
        { new IEncodable with
            member _.Encode(helpers) = helpers.encodeString value
        }

    /// <summary>Encode a char as a single character string.</summary>
    let inline char value =
        { new IEncodable with
            member _.Encode(helpers) = helpers.encodeChar value
        }

    /// <summary>Encode a Guid as a string.</summary>
    let inline guid value = value.ToString() |> string

    /// <summary>Encode a Uri as a string, keeping the original form.</summary>
    let inline uri (value: Uri) = value.OriginalString |> string

    /// <summary>Encode a float as a number.</summary>
    let inline float value =
        { new IEncodable with
            member _.Encode(helpers) = helpers.encodeDecimalNumber value
        }

    /// <summary>Encode a float32 as a number.</summary>
    let float32 (value: float32) = float (Operators.float value)

    /// <summary>Encode a decimal as a string, so no precision is lost.</summary>
    let inline decimal (value: decimal) =
        value.ToString(CultureInfo.InvariantCulture) |> string

    /// <summary>Encode <c>null</c>.</summary>
    let inline nil<'T> =
        { new IEncodable with
            member _.Encode(helpers) = helpers.encodeNull ()
        }

    /// <summary>Encode a boolean.</summary>
    let inline bool value =
        { new IEncodable with
            member _.Encode(helpers) = helpers.encodeBool value
        }

    /// <summary>Encode an object from the given properties.</summary>
    let inline object (values: seq<string * IEncodable>) =
        { new IEncodable with
            member _.Encode(helpers) =
                values
                |> Seq.map (fun (k, v) -> (k, v.Encode(helpers)))
                |> helpers.encodeObject
        }

    /// <summary>Encode an array.</summary>
    let inline array (values: IEncodable array) =
        { new IEncodable with
            member _.Encode(helpers) =
                values
                |> Array.map (fun v -> v.Encode(helpers))
                |> helpers.encodeArray
        }

    /// <summary>Encode a list as a JSON array.</summary>
    let list (values: IEncodable list) =
        { new IEncodable with
            member _.Encode(helpers) =
                values
                |> List.map (fun v -> v.Encode(helpers))
                |> helpers.encodeList
        }

    /// <summary>Encode a sequence as a JSON array.</summary>
    let seq (values: IEncodable seq) =
        { new IEncodable with
            member _.Encode(helpers) =
                values
                |> Seq.map (fun v -> v.Encode(helpers))
                |> helpers.encodeSeq
        }

    /// <summary>Encode a ResizeArray as a JSON array.</summary>
    let resizeArray (values: IEncodable ResizeArray) =
        { new IEncodable with
            member _.Encode(helpers) =
                let result = ResizeArray(values.Count)

                for v in values do
                    result.Add(v.Encode(helpers))

                helpers.encodeResizeArray result
        }

    /// <summary>Encode each element of a list with the given encoder.</summary>
    let mapList (encoder: Encoder<'a>) (values: 'a list) : IEncodable =
        values |> List.map encoder |> list

    /// <summary>Encode each element of an array with the given encoder.</summary>
    let mapArray (encoder: Encoder<'a>) (values: 'a array) : IEncodable =
        values |> Array.map encoder |> array

    /// <summary>Encode each element of a sequence with the given encoder.</summary>
    let mapSeq (encoder: Encoder<'a>) (values: 'a seq) : IEncodable =
        values |> Seq.map encoder |> seq

    /// <summary>Encode each element of a ResizeArray with the given encoder.</summary>
    let mapResizeArray
        (encoder: Encoder<'a>)
        (values: 'a ResizeArray)
        : IEncodable
        =
        values |> Seq.map encoder |> ResizeArray |> resizeArray

    /// <summary>Encode a map as an object, one property per key.</summary>
    let dict (values: Map<string, IEncodable>) : IEncodable =
        values |> Map.toSeq |> object

    /// <summary>Encode a bigint as a string.</summary>
    let inline bigint (value: bigint) = value.ToString() |> string

    /// <summary>Encode a DateTimeOffset as an ISO 8601 string.</summary>
    let inline datetimeOffset (value: DateTimeOffset) =
        value.ToString("O", CultureInfo.InvariantCulture) |> string

    /// <summary>Encode a TimeSpan as a string.</summary>
    let inline timespan value = value.ToString() |> string

    /// <summary>Encode a DateTime as an ISO 8601 string.</summary>
    let inline datetime (value: DateTime) =
        value.ToString("O", CultureInfo.InvariantCulture) |> string

    /// <summary>Encode an sbyte as a number.</summary>
    let inline sbyte (value: sbyte) =
        { new IEncodable with
            member _.Encode(helpers) =
                helpers.encodeSignedIntegralNumber (int32 value)
        }

    /// <summary>Encode a byte as a number.</summary>
    let inline byte (value: byte) =
        { new IEncodable with
            member _.Encode(helpers) =
                helpers.encodeUnsignedIntegralNumber (uint32 value)
        }

    /// <summary>Encode an int16 as a number.</summary>
    let inline int16 (value: int16) =
        { new IEncodable with
            member _.Encode(helpers) =
                helpers.encodeSignedIntegralNumber (int32 value)
        }

    /// <summary>Encode a uint16 as a number.</summary>
    let inline uint16 (value: uint16) =
        { new IEncodable with
            member _.Encode(helpers) =
                helpers.encodeUnsignedIntegralNumber (uint32 value)
        }

    /// <summary>Encode an int as a number.</summary>
    let inline int (value: int) =
        { new IEncodable with
            member _.Encode(helpers) =
                helpers.encodeSignedIntegralNumber value
        }

    /// <summary>Encode a uint32 as a number.</summary>
    let inline uint32 (value: uint32) =
        { new IEncodable with
            member _.Encode(helpers) =
                helpers.encodeUnsignedIntegralNumber value
        }

    /// <summary>Encode an int64 as a string, so no precision is lost.</summary>
    let inline int64 (value: int64) =
        value.ToString(CultureInfo.InvariantCulture) |> string

    /// <summary>Encode a uint64 as a string, so no precision is lost.</summary>
    let inline uint64 (value: uint64) =
        value.ToString(CultureInfo.InvariantCulture) |> string

    /// <summary>Encode unit as <c>null</c>.</summary>
    let inline unit () = nil

    /// <summary>Encode a tuple of 2 elements as a JSON array.</summary>
    let tuple2 (enc1: Encoder<'T1>) (enc2: Encoder<'T2>) (v1, v2) : IEncodable =
        array
            [|
                enc1 v1
                enc2 v2
            |]

    /// <summary>Encode a tuple of 3 elements as a JSON array.</summary>
    let tuple3
        (enc1: Encoder<'T1>)
        (enc2: Encoder<'T2>)
        (enc3: Encoder<'T3>)
        (v1, v2, v3)
        : IEncodable
        =
        array
            [|
                enc1 v1
                enc2 v2
                enc3 v3
            |]

    /// <summary>Encode a tuple of 4 elements as a JSON array.</summary>
    let tuple4
        (enc1: Encoder<'T1>)
        (enc2: Encoder<'T2>)
        (enc3: Encoder<'T3>)
        (enc4: Encoder<'T4>)
        (v1, v2, v3, v4)
        : IEncodable
        =
        array
            [|
                enc1 v1
                enc2 v2
                enc3 v3
                enc4 v4
            |]

    /// <summary>Encode a tuple of 5 elements as a JSON array.</summary>
    let tuple5
        (enc1: Encoder<'T1>)
        (enc2: Encoder<'T2>)
        (enc3: Encoder<'T3>)
        (enc4: Encoder<'T4>)
        (enc5: Encoder<'T5>)
        (v1, v2, v3, v4, v5)
        : IEncodable
        =
        array
            [|
                enc1 v1
                enc2 v2
                enc3 v3
                enc4 v4
                enc5 v5
            |]

    /// <summary>Encode a tuple of 6 elements as a JSON array.</summary>
    let tuple6
        (enc1: Encoder<'T1>)
        (enc2: Encoder<'T2>)
        (enc3: Encoder<'T3>)
        (enc4: Encoder<'T4>)
        (enc5: Encoder<'T5>)
        (enc6: Encoder<'T6>)
        (v1, v2, v3, v4, v5, v6)
        : IEncodable
        =
        array
            [|
                enc1 v1
                enc2 v2
                enc3 v3
                enc4 v4
                enc5 v5
                enc6 v6
            |]

    /// <summary>Encode a tuple of 7 elements as a JSON array.</summary>
    let tuple7
        (enc1: Encoder<'T1>)
        (enc2: Encoder<'T2>)
        (enc3: Encoder<'T3>)
        (enc4: Encoder<'T4>)
        (enc5: Encoder<'T5>)
        (enc6: Encoder<'T6>)
        (enc7: Encoder<'T7>)
        (v1, v2, v3, v4, v5, v6, v7)
        : IEncodable
        =
        array
            [|
                enc1 v1
                enc2 v2
                enc3 v3
                enc4 v4
                enc5 v5
                enc6 v6
                enc7 v7
            |]

    /// <summary>Encode a tuple of 8 elements as a JSON array.</summary>
    let tuple8
        (enc1: Encoder<'T1>)
        (enc2: Encoder<'T2>)
        (enc3: Encoder<'T3>)
        (enc4: Encoder<'T4>)
        (enc5: Encoder<'T5>)
        (enc6: Encoder<'T6>)
        (enc7: Encoder<'T7>)
        (enc8: Encoder<'T8>)
        (v1, v2, v3, v4, v5, v6, v7, v8)
        : IEncodable
        =
        array
            [|
                enc1 v1
                enc2 v2
                enc3 v3
                enc4 v4
                enc5 v5
                enc6 v6
                enc7 v7
                enc8 v8
            |]


    /// <summary>Encode a map as an array of <c>[ key, value ]</c> pairs. Use it when the key is not a string.</summary>
    let map
        (keyEncoder: Encoder<'key>)
        (valueEncoder: Encoder<'value>)
        (values: Map<'key, 'value>)
        : IEncodable
        =
        values
        |> Map.toList
        |> List.map (tuple2 keyEncoder valueEncoder)
        |> list

    /// <summary>Defer the construction of an encoder until it is first used, for recursive types.</summary>
    let lazily<'t> (enc: Lazy<Encoder<'t>>) : Encoder<'t> =
        fun (x: 't) -> enc.Value x

    ////////////
    /// Enum ///
    ///////////

    module Enum =

        /// <summary>Encode an enum with an underlying byte as a number.</summary>
        let byte<'TEnum when 'TEnum: enum<byte>> (value: 'TEnum) : IEncodable =
            LanguagePrimitives.EnumToValue value |> byte

        /// <summary>Encode an enum with an underlying sbyte as a number.</summary>
        let sbyte<'TEnum when 'TEnum: enum<sbyte>>
            (value: 'TEnum)
            : IEncodable
            =
            LanguagePrimitives.EnumToValue value |> sbyte

        /// <summary>Encode an enum with an underlying int16 as a number.</summary>
        let int16<'TEnum when 'TEnum: enum<int16>>
            (value: 'TEnum)
            : IEncodable
            =
            LanguagePrimitives.EnumToValue value |> int16

        /// <summary>Encode an enum with an underlying uint16 as a number.</summary>
        let uint16<'TEnum when 'TEnum: enum<uint16>>
            (value: 'TEnum)
            : IEncodable
            =
            LanguagePrimitives.EnumToValue value |> uint16

        /// <summary>Encode an enum with an underlying int as a number.</summary>
        let int<'TEnum when 'TEnum: enum<int>> (value: 'TEnum) : IEncodable =
            LanguagePrimitives.EnumToValue value |> int

        /// <summary>Encode an enum with an underlying uint32 as a number.</summary>
        let uint32<'TEnum when 'TEnum: enum<uint32>>
            (value: 'TEnum)
            : IEncodable
            =
            LanguagePrimitives.EnumToValue value |> uint32

    /// <summary>
    /// Encode <c>Some x</c> as <c>x</c>, and <c>None</c> as <c>null</c>.
    /// </summary>
    /// <remarks>
    /// Lossy: a nested option does not round-trip. Use <see cref="losslessOption"/> when you need
    /// the distinction.
    /// </remarks>
    let lossyOption (encoder: Encoder<'a>) =
        Option.map encoder >> Option.defaultWith (fun _ -> nil)

    /// <summary>
    /// Encode an option as an object carrying the case, so a nested option round-trips.
    /// </summary>
    /// <remarks>
    /// <c>Some x</c> gives <c>{ "$type": "option", "$case": "some", "$value": x }</c>, and
    /// <c>None</c> gives <c>{ "$type": "option", "$case": "none" }</c>.
    /// </remarks>
    let losslessOption (encoder: Encoder<'a>) (value: 'a option) =
        match value with
        | Some v ->
            object
                [
                    "$type", string "option"
                    "$case", string "some"
                    "$value", encoder v
                ]
        | None ->
            object
                [
                    "$type", string "option"
                    "$case", string "none"
                ]

    /// <summary>Encode a <see cref="T:Thoth.Json.Core.Json"/> value as it stands.</summary>
    let rec value: Encoder<Json> =
        fun json ->
            match json with
            | Json.Null -> nil
            | Json.Boolean b -> bool b
            | Json.String s -> string s
            | Json.Number f -> float f
            | Json.Array xs -> list (List.map value xs)
            | Json.Object kvps ->
                kvps |> Seq.map (fun (k, v) -> k, value v) |> object

    /// <summary>Write an encodable through the given helpers, giving the runtime's own JSON value.</summary>
    let inline toJsonValue
        (helpers: IEncoderHelpers<'JsonValue>)
        (json: IEncodable)
        =
        json.Encode(helpers)

    /// <summary>The encoder half of a codec.</summary>
    let codec (c: Codec<'t>) : Encoder<'t> = c.Encoder

File: assets/live-example/precompiled/92e53be0d9c69a14/packages/Thoth.Json.Core/Encode.js

import { isNegativeInfinity, isPositiveInfinity } from "fable-library-js/Double.js";
import { map as map_1 } from "fable-library-js/List.js";
import { map as map_2 } from "fable-library-js/Seq.js";
import { disposeSafe, getEnumerator } from "fable-library-js/Util.js";
import { map as map_3 } from "fable-library-js/Array.js";
import { toList, toSeq } from "fable-library-js/Map.js";
import { value as value_6 } from "fable-library-js/Option.js";

/**
 * Encode a float as a number.
 */
export function float(value_1) {
    return {
        Encode(helpers) {
            return Number.isNaN(value_1) ? helpers.encodeString("NaN") : (isPositiveInfinity(value_1) ? helpers.encodeString("Infinity") : (isNegativeInfinity(value_1) ? helpers.encodeString("-Infinity") : helpers.encodeDecimalNumber(value_1)));
        },
    };
}

/**
 * Encode a float32 as a number.
 */
export function float32(value_1) {
    return float(value_1);
}

/**
 * Encode a list as a JSON array.
 */
export function list(values) {
    return {
        Encode(helpers) {
            const arg = map_1((v) => v.Encode(helpers), values);
            return helpers.encodeList(arg);
        },
    };
}

/**
 * Encode a sequence as a JSON array.
 */
export function seq(values) {
    return {
        Encode(helpers) {
            const arg = map_2((v) => v.Encode(helpers), values);
            return helpers.encodeSeq(arg);
        },
    };
}

/**
 * Encode a ResizeArray as a JSON array.
 */
export function resizeArray(values) {
    return {
        Encode(helpers) {
            const result = [];
            let enumerator = getEnumerator(values);
            try {
                while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
                    const v = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
                    void (result.push(v.Encode(helpers)));
                }
            }
            finally {
                disposeSafe(enumerator);
            }
            return helpers.encodeResizeArray(result);
        },
    };
}

/**
 * Encode each element of a list with the given encoder.
 */
export function mapList(encoder, values) {
    return list(map_1(encoder, values));
}

/**
 * Encode each element of an array with the given encoder.
 */
export function mapArray(encoder, values) {
    const values_1 = map_3(encoder, values);
    return {
        Encode(helpers) {
            const arg = map_3((v) => v.Encode(helpers), values_1);
            return helpers.encodeArray(arg);
        },
    };
}

/**
 * Encode each element of a sequence with the given encoder.
 */
export function mapSeq(encoder, values) {
    return seq(map_2(encoder, values));
}

/**
 * Encode each element of a ResizeArray with the given encoder.
 */
export function mapResizeArray(encoder, values) {
    let collection;
    return resizeArray((collection = map_2(encoder, values), Array.from(collection)));
}

/**
 * Encode a map as an object, one property per key.
 */
export function dict(values) {
    const values_1 = toSeq(values);
    return {
        Encode(helpers) {
            const arg = map_2((tupledArg) => [tupledArg[0], tupledArg[1].Encode(helpers)], values_1);
            return helpers.encodeObject(arg);
        },
    };
}

/**
 * Encode a tuple of 2 elements as a JSON array.
 */
export function tuple2(enc1, enc2, v1, v2) {
    const values = [enc1(v1), enc2(v2)];
    return {
        Encode(helpers) {
            const arg = map_3((v) => v.Encode(helpers), values);
            return helpers.encodeArray(arg);
        },
    };
}

/**
 * Encode a tuple of 3 elements as a JSON array.
 */
export function tuple3(enc1, enc2, enc3, v1, v2, v3) {
    const values = [enc1(v1), enc2(v2), enc3(v3)];
    return {
        Encode(helpers) {
            const arg = map_3((v) => v.Encode(helpers), values);
            return helpers.encodeArray(arg);
        },
    };
}

/**
 * Encode a tuple of 4 elements as a JSON array.
 */
export function tuple4(enc1, enc2, enc3, enc4, v1, v2, v3, v4) {
    const values = [enc1(v1), enc2(v2), enc3(v3), enc4(v4)];
    return {
        Encode(helpers) {
            const arg = map_3((v) => v.Encode(helpers), values);
            return helpers.encodeArray(arg);
        },
    };
}

/**
 * Encode a tuple of 5 elements as a JSON array.
 */
export function tuple5(enc1, enc2, enc3, enc4, enc5, v1, v2, v3, v4, v5) {
    const values = [enc1(v1), enc2(v2), enc3(v3), enc4(v4), enc5(v5)];
    return {
        Encode(helpers) {
            const arg = map_3((v) => v.Encode(helpers), values);
            return helpers.encodeArray(arg);
        },
    };
}

/**
 * Encode a tuple of 6 elements as a JSON array.
 */
export function tuple6(enc1, enc2, enc3, enc4, enc5, enc6, v1, v2, v3, v4, v5, v6) {
    const values = [enc1(v1), enc2(v2), enc3(v3), enc4(v4), enc5(v5), enc6(v6)];
    return {
        Encode(helpers) {
            const arg = map_3((v) => v.Encode(helpers), values);
            return helpers.encodeArray(arg);
        },
    };
}

/**
 * Encode a tuple of 7 elements as a JSON array.
 */
export function tuple7(enc1, enc2, enc3, enc4, enc5, enc6, enc7, v1, v2, v3, v4, v5, v6, v7) {
    const values = [enc1(v1), enc2(v2), enc3(v3), enc4(v4), enc5(v5), enc6(v6), enc7(v7)];
    return {
        Encode(helpers) {
            const arg = map_3((v) => v.Encode(helpers), values);
            return helpers.encodeArray(arg);
        },
    };
}

/**
 * Encode a tuple of 8 elements as a JSON array.
 */
export function tuple8(enc1, enc2, enc3, enc4, enc5, enc6, enc7, enc8, v1, v2, v3, v4, v5, v6, v7, v8) {
    const values = [enc1(v1), enc2(v2), enc3(v3), enc4(v4), enc5(v5), enc6(v6), enc7(v7), enc8(v8)];
    return {
        Encode(helpers) {
            const arg = map_3((v) => v.Encode(helpers), values);
            return helpers.encodeArray(arg);
        },
    };
}

/**
 * Encode a map as an array of <c>[ key, value ]</c> pairs. Use it when the key is not a string.
 */
export function map(keyEncoder, valueEncoder, values) {
    return list(map_1((tupledArg) => tuple2(keyEncoder, valueEncoder, tupledArg[0], tupledArg[1]), toList(values)));
}

/**
 * Defer the construction of an encoder until it is first used, for recursive types.
 */
export function lazily(enc, x) {
    return enc.Value(x);
}

/**
 * Encode an enum with an underlying byte as a number.
 */
export function Enum_byte(value_1) {
    return {
        Encode(helpers) {
            return helpers.encodeUnsignedIntegralNumber(value_1);
        },
    };
}

/**
 * Encode an enum with an underlying sbyte as a number.
 */
export function Enum_sbyte(value_1) {
    return {
        Encode(helpers) {
            return helpers.encodeSignedIntegralNumber(value_1);
        },
    };
}

/**
 * Encode an enum with an underlying int16 as a number.
 */
export function Enum_int16(value_1) {
    return {
        Encode(helpers) {
            return helpers.encodeSignedIntegralNumber(value_1);
        },
    };
}

/**
 * Encode an enum with an underlying uint16 as a number.
 */
export function Enum_uint16(value_1) {
    return {
        Encode(helpers) {
            return helpers.encodeUnsignedIntegralNumber(value_1);
        },
    };
}

/**
 * Encode an enum with an underlying int as a number.
 */
export function Enum_int(value_1) {
    return {
        Encode(helpers) {
            return helpers.encodeSignedIntegralNumber(value_1);
        },
    };
}

/**
 * Encode an enum with an underlying uint32 as a number.
 */
export function Enum_uint32(value_1) {
    return {
        Encode(helpers) {
            return helpers.encodeUnsignedIntegralNumber(value_1);
        },
    };
}

/**
 * Encode <c>Some x</c> as <c>x</c>, and <c>None</c> as <c>null</c>.
 */
export function lossyOption(encoder) {
    return (arg) => {
        let option_3;
        const option_1 = arg;
        option_3 = ((option_1 != null) ? encoder(value_6(option_1)) : undefined);
        return (option_3 != null) ? option_3 : {
            Encode(helpers) {
                return helpers.encodeNull();
            },
        };
    };
}

/**
 * Encode an option as an object carrying the case, so a nested option round-trips.
 */
export function losslessOption(encoder, value_1) {
    if (value_1 == null) {
        const values_1 = [["$type", {
            Encode(helpers_3) {
                return helpers_3.encodeString("option");
            },
        }], ["$case", {
            Encode(helpers_4) {
                return helpers_4.encodeString("none");
            },
        }]];
        return {
            Encode(helpers_5) {
                const arg_1 = map_2((tupledArg_1) => [tupledArg_1[0], tupledArg_1[1].Encode(helpers_5)], values_1);
                return helpers_5.encodeObject(arg_1);
            },
        };
    }
    else {
        const v = value_6(value_1);
        const values = [["$type", {
            Encode(helpers) {
                return helpers.encodeString("option");
            },
        }], ["$case", {
            Encode(helpers_1) {
                return helpers_1.encodeString("some");
            },
        }], ["$value", encoder(v)]];
        return {
            Encode(helpers_2) {
                const arg = map_2((tupledArg) => [tupledArg[0], tupledArg[1].Encode(helpers_2)], values);
                return helpers_2.encodeObject(arg);
            },
        };
    }
}

/**
 * Encode a <see cref="T:Thoth.Json.Core.Json"/> value as it stands.
 */
export function value(json) {
    switch (json.tag) {
        case 3:
            return {
                Encode(helpers_1) {
                    return helpers_1.encodeBool(json.fields[0]);
                },
            };
        case 0:
            return {
                Encode(helpers_2) {
                    return helpers_2.encodeString(json.fields[0]);
                },
            };
        case 1:
            return float(json.fields[0]);
        case 5:
            return list(map_1(value, json.fields[0]));
        case 4: {
            const values = map_2((tupledArg) => [tupledArg[0], value(tupledArg[1])], json.fields[0]);
            return {
                Encode(helpers_3) {
                    const arg = map_2((tupledArg_1) => [tupledArg_1[0], tupledArg_1[1].Encode(helpers_3)], values);
                    return helpers_3.encodeObject(arg);
                },
            };
        }
        default:
            return {
                Encode(helpers) {
                    return helpers.encodeNull();
                },
            };
    }
}

/**
 * The encoder half of a codec.
 */
export function codec(c) {
    return c.Encoder;
}

